### DISCUSSION TITLE
trigger on user delete and more

### DISCUSSION THREAD

**Discussion by EdgarsLv** [supabase/supabase#6539] [Questions]
Have table conversations.
When user involved in conversation gets deleted hes  fkey column id (user_one or user_two) is set to null.

Wanted to create trigger on user delete which checks if both user id in conversations are null then delete conversation.

schema = public
return type = trigger

and code below
`begin
DELETE FROM conversations WHERE user_one = null AND user_two = null;
  RETURN NULL;
end;`

Enabling this throws error on user delete and i no longer can delete user.

1. What's wrong here?

And another problem.
I have userimages table as follow:

![image](https://user-images.githubusercontent.com/56837461/164967052-1f6b2e57-4bc8-45f0-85da-fda69f919f8c.png)

When user gets deleted i want set ON DELETE CASCADE but i get this error:
![image](https://user-images.githubusercontent.com/56837461/164967137-a5bfc2b0-9db0-4665-857a-b4979e335759.png)

I already have done this to other tables.

2. What's wrong here? Why i gett error that constraind doesn't exist?

**Comment by EdgarsLv** [comment]
So i tried like this:

![image](https://user-images.githubusercontent.com/56837461/164982015-a9cd81f9-20a2-4b83-b4ab-23034d86da09.png)

now it was success.

But now i can't fetch like this:
![image](https://user-images.githubusercontent.com/56837461/164982057-328db547-ee3b-433b-9755-47ad3d3d5847.png)

It returns nothing.

What is going og?

**Comment by GaryAustin1** [comment]
1)  If that is a delete after trigger on auth.users table you need to return new.  Returning null kills the the delete operation.

2)  In your 2nd part just asked...  If users is auth.users then you can't access that from the API.

**Reply by EdgarsLv** [reply]
What you mean with 2nd? 

I already did that to other tables and everything works fine. 
userimages was last table i had to do it.

**Reply by GaryAustin1** [reply]
2 was in regards to your asking why you could not access users, returns nothing.  If that is auth.users table you can not access from the api.

As far as why you got the constraint error with drop and then add, drop would fail if you did not already have that constraint.  When you did just add it worked.  Now if you do add it will fail, unless you do the drop first...

**Reply by EdgarsLv** [reply]
I mean i cant't fetch foreign table anymore. `.from('users')
        .select(`*, userimages (images)`)`
no error and returns nothing. 

this works.
`.from('users')
        .select(`*`)`

and about 1. returning new doesn't works. throws error when deleting user

**Reply by GaryAustin1** [reply]
Do you have a public.users table in addition to auth.users?   If so things might be messed up as you did not use public.users, auth.users

**Reply by EdgarsLv** [reply]
> Do you have a public.users table in addition to auth.users? If so things might be messed up as you did not use public.users, auth.users

I have public users table. Everythink was working fine till i run this command: 
`ALTER TABLE userimages
ADD CONSTRAINT userimages_user_id_fkey
   FOREIGN KEY (user_id)
   REFERENCES users
   ON DELETE CASCADE;`

now i can't fetch foreign tables. and there is no error or anything. I can fetch anything from users table but if i add foreign table userimages it doesn't retuns anything.

**Reply by GaryAustin1** [reply]
Can you show your user table columns names and types?

**Comment by EdgarsLv** [comment]
Deleted userimages table and created new one and now fetch works fine.

But still there is problems with that table on user delete.

All other tables on auth user delete are deleted or fkeys set to null as i wanted including storage objects. But userimmages table is preventing me from deleting user and throws error.

If i delete coresponding userimages row from dashboard, then i can delete user from app which also deletes storage objects as i want.

I was trying to set userimages ON DELETE CASCADE when auth user is deleted, but that doesn't works.
Maybe then i should set userimages ON DELETE CASCADE when storage object is deleted which will gets deleted when auth user is deleted?
